### PR TITLE
feat: platform=local

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds the ability to run with `--platform=local` to run a dry run against the local fs.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/local/index.ts
Comment (rarkins) [inline lib/config/presets/local/index.ts:27]: This means if someone runs locally against a repo with (valid) `local>` presets then it will fail

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:1]: There's lots of dummy functions here but this still seemed to be the cleanest way to achieve what I wanted

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:14]: This can still be overridden (e.g. `--dry-run=extract`)

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:16]: This is so that we don't delete the cwd at the end

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:17]: so that it runs regardless of whether there's a config

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:29]: Essentially these are all dummy values

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [inline lib/modules/platform/local/scm.ts:5]: This is ok for now because we only allow one repo per run

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [inline lib/modules/platform/local/scm.ts:36]: I needed to add these otherwise getFileList() took an eternity.

- probably should remove the hardcoded node_modules
- ideally it would automatically ignore everything in `.gitignore`
- `globby` does the above automatically, but it's on v13 with ESM and v11 was the last release to support non-ESM

File: lib/workers/global/autodiscover.ts
Comment (rarkins) [inline lib/workers/global/autodiscover.ts:24]: In theory we could support multiple repos in a single run, but it adds complexity and I'm not sure if it's really a valid or common use case?

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:111]: baseBranches makes no sense if we are not using git

**Comment by rarkins** [comment]
I can see how a git-aware local could also make sense in some scenarios, perhaps it could be called platform=local-git in future

**Comment by rarkins** [comment]
BTW this is the 3rd or 4th time I've attempted this PR. All preceding attempts were cut short by Renovate proceeding to delete my entire local clone/source unintentionally (`localDir`..)

**Comment by rarkins** [comment]
The primary use case for this is:
- People who want to test their config locally (either for extract or lookup result) without needing to "push to production"
- And for whom baseBranches doesn't matter

File: lib/modules/platform/local/index.ts
Comment (viceice) [reply lib/modules/platform/local/index.ts:29]: why not fetching the head branch of local repo?

File: lib/modules/platform/local/scm.ts
Comment (viceice) [reply lib/modules/platform/local/scm.ts:5]: you could reset in init repo, or throw there if someone tries multiple repos

File: lib/modules/platform/local/scm.ts
Comment (viceice) [reply lib/modules/platform/local/scm.ts:36]: why not call git?

File: lib/modules/platform/local/index.ts
Comment (rarkins) [reply lib/modules/platform/local/index.ts:29]: I want this to work in non git situations too

File: lib/workers/global/autodiscover.ts
Comment (viceice) [reply lib/workers/global/autodiscover.ts:24]: i think a single repo is ok for now. maybe throw or log as error?

Comment (rarkins) [reply lib/workers/global/autodiscover.ts:24]: Maybe let's abort/error if any arguments are given

Comment (viceice) [reply lib/workers/global/autodiscover.ts:24]: sounds good

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [reply lib/modules/platform/local/scm.ts:6]: Let's limit local to one repo

File: lib/modules/platform/local/index.ts
Comment (rarkins) [inline lib/modules/platform/local/index.ts:1]: This file is just boilerplate to satisfy types, and two functions return static objects

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [inline lib/modules/platform/local/scm.ts:37]:
```suggestion
        childrenIgnored: (p) => p.isNamed('.git'),
```

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [reply lib/modules/platform/local/scm.ts]: Added try/catch using git

File: lib/modules/platform/local/scm.ts
Comment (viceice) [inline lib/modules/platform/local/scm.ts:1]: needs tests before merge 🙃

File: lib/modules/platform/local/scm.ts
Comment (viceice) [inline lib/modules/platform/local/scm.ts:40]: we can later test for `.git` file or folder

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/platform/local/scm.ts
Comment (viceice) [inline lib/modules/platform/local/scm.ts:9]: looks unused?

Comment (rarkins) [inline lib/modules/platform/local/scm.ts:9]:
```suggestion
```

File: lib/modules/platform/local/scm.ts
Comment (viceice) [inline lib/modules/platform/local/scm.ts:35]: why execSync when we can be async here?

**Review by viceice — COMMENTED** [review]
otherwise LGTM.

File: lib/modules/platform/local/scm.ts
Comment (rarkins) [reply lib/modules/platform/local/scm.ts:35]: 1. I couldn't get the mocking working when it was promisified exec
2. I didn't think it matters for this use case as we block on this getFileList() anyway

Comment (viceice) [reply lib/modules/platform/local/scm.ts:35]: 😏

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
needs prettifying

**Review by secustor — DISMISSED** [review]

**Review by JamieMagee — DISMISSED** [review]

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**
